{"cells":[{"cell_type":"markdown","metadata":{},"source":["\u003cp style=\"text-align:center\"\u003e\n","    \u003ca href=\"https://skills.network/?utm_medium=Exinfluencer\u0026utm_source=Exinfluencer\u0026utm_content=000026UJ\u0026utm_term=10006555\u0026utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork1005-2022-01-01\" target=\"_blank\"\u003e\n","    \u003cimg src=\"https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png\" width=\"200\" alt=\"Skills Network Logo\"  /\u003e\n","    \u003c/a\u003e\n","\u003c/p\u003e\n"]},{"cell_type":"markdown","metadata":{},"source":["# **Exception Handling**\n","\n","Estimated time needed: **15** minutes\n","    \n","\n","## Objectives\n","\n","After completing this lab you will be able to:\n","\n","* Understand exceptions    \n","* Handle the exceptions\n"]},{"cell_type":"markdown","metadata":{},"source":["## Table of Contents\n"]},{"cell_type":"markdown","metadata":{},"source":["* What is an Exception?\n","* Exception Handling\n"]},{"cell_type":"markdown","metadata":{},"source":["----\n"]},{"cell_type":"markdown","metadata":{},"source":["## What is an Exception?\n"]},{"cell_type":"markdown","metadata":{},"source":["In this section you will learn about what an exception is and see examples of them.\n"]},{"cell_type":"markdown","metadata":{},"source":["### Definition\n"]},{"cell_type":"markdown","metadata":{},"source":["An exception is an error that occurs during the execution of code. This error causes the code to raise an exception and if not prepared to handle it will halt the execution of the code.\n"]},{"cell_type":"markdown","metadata":{},"source":["### Examples\n"]},{"cell_type":"markdown","metadata":{},"source":["Run each piece of code and observe the exception raised\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["1/0"]},{"cell_type":"markdown","metadata":{},"source":["\u003ccode\u003eZeroDivisionError\u003c/code\u003e occurs when you try to divide by zero.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["y = a + 5"]},{"cell_type":"markdown","metadata":{},"source":["\u003ccode\u003eNameError\u003c/code\u003e -- in this case, it means that you tried to use the variable a when it was not defined.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["a = [1, 2, 3]\n","a[10]"]},{"cell_type":"markdown","metadata":{},"source":["\u003ccode\u003eIndexError\u003c/code\u003e -- in this case, it occured because you tried to access data from a list using an index that does not exist for this list.\n"]},{"cell_type":"markdown","metadata":{},"source":["There are many more exceptions that are built into Python, here is a list of them https://docs.python.org/3/library/exceptions.html\n"]},{"cell_type":"markdown","metadata":{},"source":["## Exception Handling\n"]},{"cell_type":"markdown","metadata":{},"source":["In this section you will learn how to handle exceptions. You will understand how to make your program perform specified tasks instead of halting code execution when an exception is encountered.\n"]},{"cell_type":"markdown","metadata":{},"source":["### Try Except\n"]},{"cell_type":"markdown","metadata":{},"source":["A \u003ccode\u003etry except\u003c/code\u003e will allow you to execute code that might raise an exception and in the case of any exception or a specific one we can handle or catch the exception and execute specific code. This will allow us to continue the execution of our program even if there is an exception.\n","\n","Python tries to execute the code in the \u003ccode\u003etry\u003c/code\u003e block. In this case if there is any exception raised by the code in the \u003ccode\u003etry\u003c/code\u003e block, it will be caught and the code block in the \u003ccode\u003eexcept\u003c/code\u003e block will be executed. After that, the code that comes \u003cem\u003eafter\u003c/em\u003e the try except will be executed.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# potential code before try catch\n","\n","try:\n","    # code to try to execute\n","except:\n","    # code to execute if there is an exception\n","    \n","# code that will still execute if there is an exception"]},{"cell_type":"markdown","metadata":{},"source":["### Try Except Example\n"]},{"cell_type":"markdown","metadata":{},"source":["In this example we are trying to divide a number given by the user, save the outcome in the variable \u003ccode\u003ea\u003c/code\u003e, and then we would like to print the result of the operation. When taking user input and dividing a number by it there are a couple of exceptions that can be raised. For example if we divide by zero. Try running the following block of code with \u003ccode\u003eb\u003c/code\u003e as a number. An exception will only be raised if \u003ccode\u003eb\u003c/code\u003e is zero.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["a = 1\n","\n","try:\n","    b = int(input(\"Please enter a number to divide a\"))\n","    a = a/b\n","    print(\"Success a=\",a)\n","except:\n","    print(\"There was an error\")\n","        \n"]},{"cell_type":"markdown","metadata":{},"source":["### Try Except Specific\n"]},{"cell_type":"markdown","metadata":{},"source":["A specific \u003ccode\u003etry except\u003c/code\u003e allows you to catch certain exceptions and also execute certain code depending on the exception. This is useful if you do not want to deal with some exceptions and the execution should halt. It can also help you find errors in your code that you might not be aware of. Furthermore, it can help you differentiate responses to different exceptions. In this case, the code after the try except might not run depending on the error.\n"]},{"cell_type":"markdown","metadata":{},"source":["\u003cb\u003eDo not run, just to illustrate:\u003c/b\u003e\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# potential code before try catch\n","\n","try:\n","    # code to try to execute\n","except (ZeroDivisionError, NameError):\n","    # code to execute if there is an exception of the given types\n","    \n","# code that will execute if there is no exception or a one that we are handling"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# potential code before try catch\n","\n","try:\n","    # code to try to execute\n","except ZeroDivisionError:\n","    # code to execute if there is a ZeroDivisionError\n","except NameError:\n","    # code to execute if there is a NameError\n","    \n","# code that will execute if there is no exception or a one that we are handling"]},{"cell_type":"markdown","metadata":{},"source":["You can also have an empty \u003ccode\u003eexcept\u003c/code\u003e at the end to catch an unexpected exception:\n"]},{"cell_type":"markdown","metadata":{},"source":["\u003cb\u003eDo not run, just to illustrate:\u003c/b\u003e\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# potential code before try catch\n","\n","try:\n","    # code to try to execute\n","except ZeroDivisionError:\n","    # code to execute if there is a ZeroDivisionError\n","except NameError:\n","    # code to execute if there is a NameError\n","except:\n","    # code to execute if ther is any exception\n","    \n","# code that will execute if there is no exception or a one that we are handling"]},{"cell_type":"markdown","metadata":{},"source":["### Try Except Specific Example\n"]},{"cell_type":"markdown","metadata":{},"source":["This is the same example as above, but now we will add differentiated messages depending on the exception, letting the user know what is wrong with the input.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["a = 1\n","\n","try:\n","    b = int(input(\"Please enter a number to divide a\"))\n","    a = a/b\n","    print(\"Success a=\",a)\n","except ZeroDivisionError:\n","    print(\"The number you provided cant divide 1 because it is 0\")\n","except ValueError:\n","    print(\"You did not provide a number\")\n","except:\n","    print(\"Something went wrong\")\n","        \n"]},{"cell_type":"markdown","metadata":{},"source":["### Try Except Else and Finally\n"]},{"cell_type":"markdown","metadata":{},"source":["\u003ccode\u003eelse\u003c/code\u003e allows one to check if there was no exception when executing the try block. This is useful when we want to execute something only if there were no errors.\n"]},{"cell_type":"markdown","metadata":{},"source":["\u003cb\u003edo not run, just to illustrate\u003c/b\u003e\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# potential code before try catch\n","\n","try:\n","    # code to try to execute\n","except ZeroDivisionError:\n","    # code to execute if there is a ZeroDivisionError\n","except NameError:\n","    # code to execute if there is a NameError\n","except:\n","    # code to execute if ther is any exception\n","else:\n","    # code to execute if there is no exception\n","    \n","# code that will execute if there is no exception or a one that we are handling"]},{"cell_type":"markdown","metadata":{},"source":["\u003ccode\u003efinally\u003c/code\u003e allows us to always execute something even if there is an exception or not. This is usually used to signify the end of the try except.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# potential code before try catch\n","\n","try:\n","    # code to try to execute\n","except ZeroDivisionError:\n","    # code to execute if there is a ZeroDivisionError\n","except NameError:\n","    # code to execute if there is a NameError\n","except:\n","    # code to execute if ther is any exception\n","else:\n","    # code to execute if there is no exception\n","finally:\n","    # code to execute at the end of the try except no matter what\n","    \n","# code that will execute if there is no exception or a one that we are handling"]},{"cell_type":"markdown","metadata":{},"source":["### Try Except Else and Finally Example\n"]},{"cell_type":"markdown","metadata":{},"source":["You might have noticed that even if there is an error the value of \u003ccode\u003ea\u003c/code\u003e is always printed. Let's use the \u003ccode\u003eelse\u003c/code\u003e and print the value of \u003ccode\u003ea\u003c/code\u003e only if there is no error.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["a = 1\n","\n","try:\n","    b = int(input(\"Please enter a number to divide a\"))\n","    a = a/b\n","except ZeroDivisionError:\n","    print(\"The number you provided cant divide 1 because it is 0\")\n","except ValueError:\n","    print(\"You did not provide a number\")\n","except:\n","    print(\"Something went wrong\")\n","else:\n","    print(\"success a=\",a)"]},{"cell_type":"markdown","metadata":{},"source":["Now lets let the user know that we are done processing their answer. Using the \u003ccode\u003efinally\u003c/code\u003e, let's add a print.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["a = 1\n","\n","try:\n","    b = int(input(\"Please enter a number to divide a\"))\n","    a = a/b\n","except ZeroDivisionError:\n","    print(\"The number you provided cant divide 1 because it is 0\")\n","except ValueError:\n","    print(\"You did not provide a number\")\n","except:\n","    print(\"Something went wrong\")\n","else:\n","    print(\"success a=\",a)\n","finally:\n","    print(\"Processing Complete\")"]},{"cell_type":"markdown","metadata":{},"source":["## Authors\n"]},{"cell_type":"markdown","metadata":{},"source":["\u003ca href=\"https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer\u0026utm_source=Exinfluencer\u0026utm_content=000026UJ\u0026utm_term=10006555\u0026utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork1005-2022-01-01\" target=\"_blank\"\u003eJoseph Santarcangelo\u003c/a\u003e\n"]},{"cell_type":"markdown","metadata":{},"source":["## Change Log\n","\n","|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |\n","|---|---|---|---|\n","| 2020-09-02  | 2.0  | Simran | Template updates to the file|\n","|   |   |   |   |\n","|   |   |   |   |\n","\n","\n","\n","\n","## \u003ch3 align=\"center\"\u003e © IBM Corporation 2020. All rights reserved. \u003ch3/\u003e\n"]}],"metadata":{"kernelspec":{"display_name":"Python","language":"python","name":"conda-env-python-py"},"language_info":{"name":""}},"nbformat":4,"nbformat_minor":4}